demes/__main__.py

```diff
@@ -0,0 +1,168 @@
+import io
+import sys
+import itertools
+import argparse
+from typing import Iterator, Tuple, Union
+
+import demes
+from . import ms
+
+
+class ParseCommand:
+    def __init__(self, subparsers):
+        parser = subparsers.add_parser(
+            "parse", help="Parse models and output canonical forms."
+        )
+        parser.set_defaults(func=self)
+        parser.add_argument(
+            "-j",
+            "--json",
+            action="store_true",
+            default=False,
+            help="Output a JSON-formatted model. YAML is output by default.",
+        )
+        parser.add_argument(
+            "-s",
+            "--simplified",
+            action="store_true",
+            default=False,
+            help=(
+                "Output a simplified model. A fully-qualified model "
+                "is output by default."
+            ),
+        )
+        parser.add_argument(
+            "filename",
+            type=argparse.FileType(),
+            help=(
+                "Filename of the model. The special value '-' may be used to "
+                "read from stdin."
+            ),
+        )
+
+    def __call__(self, args: argparse.Namespace) -> None:
+        """
+        Dump JSON/YAML output in simplified/fully-qualified variants.
+        """
+        if args.json:
+            output_format = "json"
+        else:
+            output_format = "yaml"
+
+        num_documents, graphs = self.load_and_count_documents(args.filename)
+        if num_documents == 0:
+            # Input file is empty.
+            pass
+        elif num_documents == 1:
+            demes.dump(
+                next(graphs),
+                sys.stdout,
+                simplified=args.simplified,
+                format=output_format,
+            )
+        else:
+            if output_format != "yaml":
+                raise RuntimeError(
+                    "multi-document files only supported with YAML output",
+                )
+            demes.dump_all(graphs, sys.stdout, simplified=args.simplified)
+
+    def load_and_count_documents(
+        self, filename: Union[str, io.TextIOBase]
+    ) -> Tuple[int, Iterator[demes.Graph]]:
+        """
+        Count the documents in the file, returning the count and an iterator
+        over the graphs. The returned document count is:
+            0 for zero documents,
+            1 for one document, and
+            2 for two or more documents.
+        """
+        graph_generator = demes.load_all(filename)
```

Comment:
Here maybe? Something to do with the generator not being unwound when the error is thrown? Seems unlikely tbh.